Port StereoSIDPlayer/SIDplay to Unix / macOS

pda opened the pull request:
Port SIDplay to work on macOS and other unix/posix systems using `termios.h` for serial communication.

Works for me on macOS 11.5.2. The Windows code paths should be unchanged behind `#ifdef _WIN32`, but I don't have a Windows GCC environment handy to test it.

----

_Original pull request message:_

_This is work in progress porting the Windows component of StereoSIDPlayer to macOS and hopefully other unix-like systems._

_It currently compiles, connects to the ARMSID, and produces sound, but it's completely garbled. Maybe a timing issue, or something not quite right about the serial port configuration._

_I haven't preserved the Windows code in this spike, but if/when the unix version works, the Windows code can be brought back in `ifdef` blocks, perhaps after a bit of refactoring/abstracting._

_Unfortunately all the line endings have changed from windows `\r\n` to unix `\n`, so you may need to [ignore whitespace changes](https://github.com/nobomi/Arduino-ARMSID-configurator/pull/2/files?w=1) to see what's really changed._

_I have no idea if/when I'll have time to work more on this, but I'll open it as a draft PR just in case it's useful._

pda commented:
I've ben banging my head against this a bit.
I tested the precompiled `SIDplay410.exe` on a Windows machine with this same Arduino/ARMSID and it works perfectly.
But this macOS version sounds like a broken dialup modem.

I added a bit more debugging…
- `[` instead of `o` (Xon)
- `]` instead of `f` (Xoff)
- reporting the number of empty reads waiting for Xon.
- reporting the number of writes between Xon and Xoff.

```
( 3875456 empty reads) [...........] (511 writes, 2044 bytes)
( 1967228 empty reads) [................] (731 writes, 2924 bytes)
( 4483566 empty reads) [...........] (559 writes, 2236 bytes)
( 1506778 empty reads) [....................] (908 writes, 3632 bytes)
( 2668420 empty reads) [................] (759 writes, 3036 bytes)
( 4531772 empty reads) [..................] (810 writes, 3240 bytes)
( 5447180 empty reads) [.................] (789 writes, 3156 bytes)
```

There's quite a long wait (about 1 second) between the `]` Xoff and the `[` Xon while those empty `read()` calls are happening.

Also, I notice that ~3000 bytes are sent before Xoff is received, whereas the Arduino code only has a 1,024 byte buffer and should send Xoff when it's half full. Maybe the Xoff wasn't coming through fast enough and the buffer was overflowing? Maybe I could add some overflow detection on the Arduino side, if it's not there already.

With a very short `nanosleep()` after each write, it looks like this:

```
( 4854040 empty reads) [...] (171 writes, 684 bytes)
( 3192920 empty reads) [.....] (203 writes, 812 bytes)
( 3847252 empty reads) [....] (204 writes, 816 bytes)
( 3141248 empty reads) [....] (174 writes, 696 bytes)
( 2235636 empty reads) [....] (206 writes, 824 bytes)
( 1810916 empty reads) [.....] (209 writes, 836 bytes)
( 2010412 empty reads) [....] (176 writes, 704 bytes)
```

684 bytes is much closer to the Ardunio/ARMSID 512/1024 byte buffer. But the SID sound output still sounds like a broken modem, and there's still heaps of empty reads between Xoff and Xon. 🤷 

I'll try to take another look soon.

nobomi commented:
Thank you for this work.
About your issues, just two ideas ...

1) be sure your serial data stream is not disrupted by any possible system filtering for example by cfmakeraw() funcion ? And line settings is 115200Bd, 8bit, 1stopbit, no parity (it seems there is no problem here)

2) all data from write command must be physically delivered before the next read (or at least before a next write), otherwise handshaking may not work. In original code it's done by the serial port timeouts settings.

The write command needs to be executed immediately, there is no time for system data collecting. And, the most important thing, then you have to wait at least 100 us for every sended byte to sending finish.
Something like a ...
```
bytes_written = write(hSerial, data, len);
if (bytes_written>0) {
   struct timespec tim, tim2;
   tim.tv_sec = 0;
   tim.tv_nsec = 100000*len;
   nanosleep(&tim , &tim2);
}
```

pda commented:
Thanks for the tips @nobomi!

> be sure your serial data stream is not disrupted by any possible system filtering for example by cfmakeraw() funcion ? And line settings is 115200Bd, 8bit, 1stopbit, no parity (it seems there is no problem here)

Yeah, I think that's all as it should be. I couldn't find a great way to get all the info about `struct termios`, so I wrote a basic dump function¹ which shows me this:

```
struct termios {
  tcflag_t c_iflag: 0x00000001 IGNBRK
  tcflag_t c_oflag: 0x00000000
  tcflag_t c_cflag: 0x00004B00 CS8 CREAD HUPCL
  tcflag_t c_lflag: 0x00000000
  cc_t c_cc[20]: VMIN:0 VTIME:0
  speed_t c_ispeed: 115200
  speed_t c_ospeed: 115200
}
```

Interestingly `cfmakeraw` doesn't change anything other than setting `IGNBRK` on the Input Options… looks like the default struct termios for a serial port is already raw.



> all data from write command must be physically delivered before the next read (or at least before a next write), otherwise handshaking may not work. In original code it's done by the serial port timeouts settings.

Yeah, I've never ported Windows serial code to macOS/unix (never ported anything, actually), and I was surprised how simple most of the changes were because of equivalent functions. But the one thing without an equivalent is `WriteTotalTimeoutConstant` and `WriteTotalTimeoutMultiplier`.

Anyway — I'll keep poking the timing and see what I can figure out. Thanks again for the tips.

----

¹ My `dump_termios`, it's not exhaustive but was helpful for this project:
```c
void dump_termios(struct termios * t) {
  fprintf(stderr, "struct termios {\n");
  fprintf(stderr, "  tcflag_t c_iflag: 0x%08lX", t->c_iflag);
  if (t->c_iflag & IGNBRK)     fprintf(stderr, " IGNBRK");
  if (t->c_iflag & IXON)     fprintf(stderr, " IXON");
  fprintf(stderr, "\n");
  fprintf(stderr, "  tcflag_t c_oflag: 0x%08lX\n", t->c_oflag);
  fprintf(stderr, "  tcflag_t c_cflag: 0x%08lX", t->c_cflag);
  switch (t->c_cflag & CSIZE) {
    case CS5: fprintf(stderr, " CS5");
    case CS6: fprintf(stderr, " CS6");
    case CS7: fprintf(stderr, " CS7");
    case CS8: fprintf(stderr, " CS8");
  }
  if (t->c_cflag & CSTOPB)     fprintf(stderr, " CSTOPB");
  if (t->c_cflag & CREAD)      fprintf(stderr, " CREAD");
  if (t->c_cflag & PARENB)     fprintf(stderr, " PARENB");
  if (t->c_cflag & PARODD)     fprintf(stderr, " PARODD");
  if (t->c_cflag & HUPCL)      fprintf(stderr, " HUPCL");
  if (t->c_cflag & CLOCAL)     fprintf(stderr, " CLOCAL");
  if (t->c_cflag & CCTS_OFLOW) fprintf(stderr, " CCTS_OFLOW");
  if (t->c_cflag & CRTSCTS)    fprintf(stderr, " CRTSCTS");
  if (t->c_cflag & CRTS_IFLOW) fprintf(stderr, " CRTS_IFLOW");
  if (t->c_cflag & MDMBUF)     fprintf(stderr, " MDMBUF");
  fprintf(stderr, "\n");
  fprintf(stderr, "  tcflag_t c_lflag: 0x%08lX\n", t->c_lflag);
  fprintf(stderr, "  cc_t c_cc[%zu]: VMIN:%u VTIME:%u\n", sizeof(t->c_cc), t->c_cc[VMIN], t->c_cc[VTIME]);
  fprintf(stderr, "  speed_t c_ispeed: %lu\n", t->c_ispeed);
  fprintf(stderr, "  speed_t c_ospeed: %lu\n", t->c_ospeed);
  fprintf(stderr, "}\n");
}
```

pda commented:
> all data from write command must be physically delivered before the next read (or at least before a next write)

Got it!

> `tcdrain()` waits until all output written to the object referred to by fd has been transmitted.
— https://linux.die.net/man/3/tcdrain

`tcdrain(hSerial)` after `write(hSerial, data, len)` gets it working 🎶🎶🎶

Now I need to tidy this up and bring back the Windows code paths…

pda committed:
SIDplay ported macOS/unix/linux (termios.h)

pda committed:
SIDplay: aim for 10us sleep after write() (macOS/unix)

Without this, the sample Hexadecimal_2SID.sid would play 99% good, but
stall for a second a few times during the track. And setting it to 100us
caused the track to play too slow. 10us seems to work nicely.

pda commented:
Done! I've updated the pull request description to say:

> Port SIDplay to work on macOS and other unix/POSIX systems using `termios.h` for serial communication.
> 
> Works for me on macOS 11.5.2. The Windows code paths should be unchanged behind `#ifdef _WIN32`, but I don't have a Windows GCC environment handy to test it.

nobomi commented:
Great work ! Using tcdrain() is a good idea. I have to test if it works even with mingw and windows.
